RFC 0005-integer_overflow_systematic_checks (#64)
rdkcentral/tasecureapi: docs/rfcs/accepted/0005-integer_overflow_systematic_checks/README.md

- Feature Name: Integer Overflow Systematic Checks
- Author(s): Bastien Simondi
- Start Date: 2023-04-07
- RFC PR: 0005
- Leader(s): Eric Berry

## Introduction

Add systematic integer overflow sanitization checks to harden the code logic and mitigate abuses.

## Motivation / use-cases

Integer overflows is a class of security vulnerabilities that is hard to avoid in complex code
bases, and is challenging for static analysers. The arithmetic behind integer overflow is well
documented over internet. All parameters and field coming from the outside world (outside world
being any code that calls inside of the TASECUREAPI code) could be of any value and should be
checked against valid ranges, but that alone falls short. All resulting value of an operation
involving one or many fields from the outside world should also be evaluated to be in within
acceptable ranges.

## Updates/Obsoletes

> -

Reviewer: Done

## Affected platforms

> -

Reviewer: Done

## Open Source Dependencies

-

## Detailed design

Systematically applies to integer operations all that involve directly or indirectly any

> field/parameter that can be under the influence of the outside world. add (A+B), sub (A-B), and mult

Reviewer 2: ```suggestion
field/parameter that can be under the influence of the outside world. add (A+B), subtract (A-B), and multiply
```

(A*B) operations are in scope. More complex operations shall be decomposed and checks for intermediate

> values, e.g. (A+B+C+B) should be decomposed on

Reviewer: Done

> - check independently A, B, C and D values

Reviewer: Done

- check for A + B
- check for (A + B) + C
- check for ((A + B) + C) + D
Same applies for any combination of add, sub and/or mult

Integer overflows sanitization checks should also apply to pointers and special care should be taking
that the resulting pointer of an operation is within expected ranges.

There are various ways of checking for integer overflows, which are documented over internet.
Nonetheless I would encourage the code owners to look at builtins macros offered by the compilers.
For instance, GCC offers Builtins, as documented at:

> https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html

Reviewer 2: ```suggestion
[GCC Integer Overflow Builtins](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html)
```

Some OS, like OP-TEE, offer Macros over the compilers builtins, see this is another way of increasing
portability that you can get inspiration from, as it can be seen in

> https://github.com/OP-TEE/optee_os/blob/master/lib/libutils/ext/include/util.h

Reviewer 2: ```suggestion
[OP-TEE Overflow Macros](https://github.com/OP-TEE/optee_os/blob/master/lib/libutils/ext/include/util.h).
```

## Drawbacks

-

## Alternatives considered

> -

Reviewer 2: ```suggestion
- None.

```

## Unresolved questions

-

## Future possibilities

-

## References

> -

Reviewer 2: ```suggestion
- [Integer Overflow or Wraparound](https://cwe.mitre.org/data/definitions/190.html).
```
